### PR TITLE
Add Partial Flashing documentation

### PR DESCRIPTION
**Pull request by microbit-sam**
Initial docs for PartialFlashingService

I've placed them in `./docs/` for now

### CODE REVIEW

File: docs/MemoryMap.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+Note: there is also a BLE Service named MemoryMapService, which is used to communicate the application's structure over BLE and not the layout of the entire flash storage.
+
+# Memory layout
+
+| Address | Description |
+|---------|-------------|
+| 7F000 | bootloader_settings_page [settings page](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v15.0.0%2Flib_bootloader.html) |
+| 7E000 | mbr_params_page / bootloader settings backup [settings page](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v15.0.0%2Flib_bootloader.html) |
+| 77000 | [Bootloader](https://github.com/microbit-foundation/v2-bootloader). [1] |
+| 75000 | Flash Data Storage (BLE Bonds + ?) (defined by us, may change depending on implemenation) |
```
Comment (jaustin) [inline docs/MemoryMap.md:14]: I'd like to be a bit more specific about what we use this bit for, as it's defined by us. Is it the bootloader/SD that touches this. Am I right in thinking that the location is defined by us, but the purpose is for SoftDevice to use?

File: docs/MemoryMap.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Memory Map
+
+This file gives an overview of the flash layout used by the nRF52833 on the micro:bit v2.
```
Comment (jaustin) [inline docs/MemoryMap.md:3]: Worth putting in a tag or commit reference to when this is written. EG 'as of v2.whatever'

File: docs/MemoryMap.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+| 77000 | [Bootloader](https://github.com/microbit-foundation/v2-bootloader). [1] |
+| 75000 | Flash Data Storage (BLE Bonds + ?) (defined by us, may change depending on implemenation) |
+| 74000 | MicroBitStorage  (is where MicroBitStorage puts the things it stores) |
+| 73000 | CODAL scratch page (is used as temporary scratch by MicroBitFlash, MicroBitFileSystem and MicroBitStorage) |
+| 1C000 | Application |
+| 01000 | SoftDevice (S113)|
+| 00000 | [Master Boot Record](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fsds_s140%2FSDS%2Fs1xx%2Fmbr_bootloader%2Fmbr_bootloader.html) |
+
+## Notes
+[1] Our [current bootloader](https://github.com/microbit-foundation/codal-microbit/blob/master/lib/bootloader.o) requires 24.3 KB. No longer reserving additional space for USB DFU
```
Comment (jaustin) [inline docs/MemoryMap.md:22]: I think it's unlikely USB DFU is how we expand this, more likely we want to add, say, display code, or sound-based errors, or... I don't know - but basically more generic 'future expansion' so we're not trapped into changing the whole memory map if the bootloader has to grow.

File: docs/bluetooth/MicroBitPartialFlashing.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+The Partial Flashing Service allows a BLE client to connect to a micro:bit and read and write the information required to partially update the firmware. For MakeCode programs this is the user's script, and for Python programs this is the file system.
+
+This service exists in partnership with the DFU service that allows a client to fully flash the micro:bit.
+
+Diagrams showing the flow of data during partial flashing can be found at [the bottom of the document](#partial flashing client).
+
+## Bluetooth Service Specification
+
+TODO: Does BLE profile exist for v2?
```
Comment (jaustin) [inline docs/bluetooth/MicroBitPartialFlashing.md:13]: For V2 we're confirming to the same profile spec as we did for V1, so linking to same place is fine. We haven't got an updated version.
Do we need to modify that spec to update it for the small variation between V1 and V2?

File: docs/bluetooth/MicroBitPartialFlashing.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+## Enabling the service
+
+Currently partial flashing is supported by programs built using MakeCode or the Python editor. This is because the service requires the user program to be appended onto the end of CODAL, something which can't be guaranteed in a pure C++ program. The partial flashing service is enabled by default for MakeCode and Python programs.
```
Comment (jaustin) [inline docs/bluetooth/MicroBitPartialFlashing.md:17]: Do we need to specify that this is the build configuarion of those projects that does this? IE, should we say "Both MakeCode and MicroPython builds currently enable this service by default"

I think it's worth explaining that this service is brought up in the uBit object based on entering pairing mode (assuming I'm right about that). Perhaps the most useful context here is that for both those languages, the service enters the system via the uBit object?

Something like
```
In order to work properly, the partial flashing service requires certain things tobe true about the memory map, and it must be possible for the MemoryMapService to be able to build a model of the flash layout. Most significantly, the service is ideal for updating data outside the main C/C++ build containing CODAL. This is currently possible for MakeCode and MicroPython builds, but not trivial for CODAL programs, where the program is linked into a single executable.

The partial flashing service is included by default in MakeCode builds for micro:bit V1 (using microbit-dal, not CODAL) and V2, and MicroPython builds for micro:bit V2 only. The Partial Flashing service is initialised in pairing mode by default, which is implemented as part of the uBit object
```
Perhaps a link to the Pairing Mode check is good? And then is the PartialFlashingService also enabled by default if Bluetooth is enabled? If so what are the config options that are relevant?

File: docs/bluetooth/MicroBitPartialFlashing.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+## Characteristic Commands
+The service consists of a single Bluetooth GATT characteristic that responds to a client's WRITE WITHOUT RESPONSE request with a BLE Notification. The characteristic supports commands to: Read the REGION INFO i.e. request Memory Map information, WRITE DATA to the flash, inform the micro:bit that the sevice has reached END OF TRANSMISSION, obtain the MICROBIT STATUS (PFS version #, current m:b mode), and to RESET the micro:bit into either application (MakeCode) or BLE mode.
```
Comment (jaustin) [inline docs/bluetooth/MicroBitPartialFlashing.md:20]: Could this be a list please?

**Review by jaustin — DISMISSED** [review]
Hey @microbit-sam I've not reviewed the specfici details of the protocol itself and the expected responses (IE the second half of MicroBitPartialFlashing.md) . Perhaps @martinwork would be happy to look over those? I've given a few general comments on the docs though that hopefully are just small tweaks

File: docs/MemoryMap.md (hunk 1 above)
Comment (martinwork) [reply docs/MemoryMap.md:14]: Two pages is the minimum size for FDS, which is a Nordic SDK component similar to MicroBitStorage or MicroBitFileSystem. The application's Nordic BLE stack, under MicroBitBLEManager, stores bond information in FDS, which uses SoftDevice to actually write to flash. The bootloader doesn't touch FDS. CODAL currently doesn't store anything else in FDS.

File: docs/bluetooth/MicroBitPartialFlashing.md (hunk 5 above)
Comment (martinwork) [reply docs/bluetooth/MicroBitPartialFlashing.md:17]: When configured, the PF service is present in both application and pairing modes.

**Comment by martinwork** [comment]
Section "Write Data Command" doesn't seem to spell out how the first and second packet offsets are combined.

File: docs/bluetooth/MicroBitPartialFlashing.md (hunk 4 above)
Comment (microbit-sam) [reply docs/bluetooth/MicroBitPartialFlashing.md:13]: The DFU service has changed to the up to date Nordic DFU

File: docs/bluetooth/MicroBitPartialFlashing.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
+# Partial Flashing Service
+
+## Introduction
+
+The Partial Flashing Service allows a BLE client to connect to a micro:bit and read and write the information required to partially update the firmware. For MakeCode programs this is the user's script, and for Python programs this is the file system.
```
Comment (jaustin) [inline docs/bluetooth/MicroBitPartialFlashing.md:5]: s/firmware/flash contents/
```suggestion
The Partial Flashing Service allows a BLE client to connect to a micro:bit and read and write the information required to partially update the flash contents. For MakeCode programs this usually means updating the user's script (compiled to asm), and for Python programs this is the entire file system.
```

File: docs/bluetooth/MicroBitPartialFlashing.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+This service exists in partnership with the DFU service that allows a client to fully flash the micro:bit.
```
Comment (jaustin) [inline docs/bluetooth/MicroBitPartialFlashing.md:7]:
```suggestion
This service exists in partnership with the DFU service that allows a client to fully flash the application storage of the micro:bit (for example, for switching between a MicroPython programme and a MakeCode programme).
```

File: docs/bluetooth/MicroBitPartialFlashing.md [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+## Bluetooth Service Specification
+
+The micro:bit v2 follows the same BLE specification as the micro:bit v1. The specification can be found [here](https://lancaster-university.github.io/microbit-docs/resources/bluetooth/bluetooth_profile.html)
+
+## Enabling the service
+
+In order to work properly, the partial flashing service requires certain things to be true about the memory map, and it must be possible for the MemoryMapService to be able to build a model of the flash layout. Most significantly, the service is ideal for updating data outside the main C/C++ build containing CODAL. This is currently possible for MakeCode and MicroPython builds, but not trivial for CODAL programs, where the program is linked into a single executable.
+
+The partial flashing service is included by default in MakeCode builds for micro:bit V1 (using microbit-dal, not CODAL) and V2, and MicroPython builds for micro:bit V2 only. The Partial Flashing service is available in both application and pairing modes.
```
Comment (jaustin) [inline docs/bluetooth/MicroBitPartialFlashing.md:19]:
```suggestion
The partial flashing service is included by default in MakeCode builds for micro:bit V1 (using microbit-dal, not CODAL) and V2, and MicroPython builds for micro:bit V2 only. The Partial Flashing service is available in both application and pairing modes.

- MakeCode: Enabled V1 (microbit-dal) and V2 (CODAL)
- MicroPython: Enabled on V2 only (CODAL)

```

**Review by jaustin — COMMENTED** [review]
Thanks Sam,

Once again, I've focused on the pre-ampble. Just minor bits to clarify. If you're happy to take these then I'm happy for this to be merged!

**Comment by microbit-sam** [comment]
@finneyj added these to a `./docs` folder for the time being

**Pull request merged**